all: use tenant-name to start an external SQL server

cthumuluru-crdb opened the pull request:
Add support for starting external SQL server using tenant name instead of tenant ID

Epic: CRDB-39093
Informs: #105340
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143447)
<!-- Reviewable:end -->

cthumuluru-crdb committed:
roachpb: add TenantIdentity to wrap both tenant ID and name

Currently, TenantID is used to generate tenant client certs for node-node
communication. It is also used to bind an a SQL server external process to
a tenant.

Our goal is to move away from TenantIDs from client interfaces to
TenantNames. To do so, we have to change `cert create-tenant-client` and
`mt start-sql` commands to accept TenantName instead of TenantIDs.

During SQL server initialization we have to resolve TenantID. But to do so
we have to change the authentication surfaces to honor client certs with
TenantNames. Once TenantName is resolved to TenantID, use it for authorization.

This PR introduces `TenantIdentity` which is meant to be the tenant security
context provided though certs. It can be either TenantID or TenantName. Use of
`TenantID` as `TenantIdentity` keeps the code backward compatible.
`TenantIdentity` will replace `TenantID` usage in authentication layer.

Epic: CRDB-39093
Informs: #105340
Release note: None

stevendanna commented on pkg/roachpb/tenant.go:
```diff
@@ -223,3 +224,83 @@ func (c *TenantNameContainer) Get() TenantName {
+	IsSystem() bool
+	// IsEqual returns whether this tenant identity is equal to another.
+	IsEqual(o TenantIdentity) bool
+	// ToString returns a string representation of the tenant identity.
+	ToString() string
```
If you make this `String()` rather than `ToString()` then this interface is also a fmt.Stringer which can be useful https://pkg.go.dev/fmt#Stringer

stevendanna reviewed (commented):
I've left a few comments. But to be honest I find it a little hard to know if this interface is a good one without seeing a few more of the next steps.  Since this doesn't abstract over names vs IDs at the point of comparison, in some ways is obfuscates things a little bit.  I'm not opposed to this going in if it helps you move forward on other work, but I also wouldn't be opposed to seeing this as the first commit in a multi-commit PR with a bit more usage code.

cthumuluru-crdb committed:
fixup! roachpb: add TenantIdentity to wrap both tenant ID and name

cthumuluru-crdb commented on pkg/roachpb/tenant.go:
```diff
@@ -223,3 +224,83 @@ func (c *TenantNameContainer) Get() TenantName {
+	// IsSet returns whether this tenant identity is set or not.
+	IsSet() bool
+	// IsValid returns an error if the tenant identity is invalid.
+	IsValid() error
```
Yes, I agree it's weird that `IsValid()` is returning an error but unfortunately `TenantName` already had it. I renamed it now but the renames touches a couple if files unrelated to this code change.

cthumuluru-crdb commented on pkg/roachpb/tenant.go:
```diff
@@ -223,3 +224,83 @@ func (c *TenantNameContainer) Get() TenantName {
+	return fmt.Sprintf("%d", t.InternalValue)
+}
+
+func (t TenantID) IsEqual(o TenantIdentity) bool {
```
Next set of comments make it clear on how it is used but for now I removed it.

cthumuluru-crdb commented:
> I've left a few comments. But to be honest I find it a little hard to know if this interface is a good one without seeing a few more of the next steps. 

Ack. Let me push a couple of other commit and we can discuss once you have a chance to see them.

>  I'm not opposed to this going in if it helps you move forward on other work, but I also wouldn't be opposed to seeing this as the first commit in a multi-commit PR with a bit more usage code.

I don't plan to commit these changes just yet. I'll wait for the other commits too before committing the PR. At a high level, the plan is:

- `TenantIdentity` is user provided tenant identity (ID/name) as security context. 
- Authentication layer uses `TenantIdentity` (ID/name) security context to authenticate.
- The code identified user provided identity as security identity. 
- Once authentication is done, TenantID is used for authorization. 

It is not possible to abstract all the call sites but the goal is to make a good effort to abstract it as much as possible.

cthumuluru-crdb committed:
all: generate tenant client certs using tenant names
`cockroach cert create-tenant-client` command is used to generate client tenant
certs. The generated cert file names are in `client-tenant.NN.crt` format where
NN is the tenant ID.

Client tenant certs are (ref: #71248):
- used as client cert for node-to-node comms (SQL server to SQL server)
- used as server cert for node-to-node comms (SQL server to SQL server)
- used as client cert for conns to KV nodes
- used as server cert for SQL clients
- used as server cert for HTTP, if ui.crt doesn't exist

`cockroach cert create-tenant-client` command takes tenant ID as an argument
followed by a variable list of hosts to be added to SAN. To stay backward
compatible and to allow generating the certs using tenant names, this PR
introduces a new flag `tenant-name-id` to control if <tenant-id> argument
should be treated as tenant ID or tenant name. When `tenant-name-id` flag
is supplied, <tenant-id> argument is treated as tenant name. Otherwise it
is treated as tenant ID (backward compatible).

---

Sample command to generate tenant client certs using tenant name:
```
$COCKROACH_HOME/cockroach cert create-tenant-client \
    --certs-dir=$HOME/.cockroach-certs \
    --ca-key=$HOME/.cockroach-certs/ca-key.pem \
    --tenant-name-id \
    "test" "127.0.0.1" "localhost"
```

The above command generates cert file (`client-tenant.test.crt`) with tenant
name in it. Tenant name is also used in CN of the certificate:
```
└─▪ openssl x509 -in ~/.cockroach-certs/client-tenant.test.crt  -text | grep test
        Subject: O=Cockroach, OU=Tenants, CN=test
```

---

Command is backward compatible when `tenant-name-id` flag is not supplied.
Sample command to generate client certs using tenant ID:
```
$COCKROACH_HOME/cockroach cert create-tenant-client \
    --certs-dir=$HOME/.cockroach-certs \
    --ca-key=$HOME/.cockroach-certs/ca-key.pem \
    999 "127.0.0.1" "localhost"
```

The above command generates cert file (`client-tenant.999.crt`) with tenant ID
in it. Tenant ID is also used in CN of the certificate:
```
└─▪ openssl x509 -in ~/.cockroach-certs/client-tenant.999.crt -text | grep 999
        Subject: O=Cockroach, OU=Tenants, CN=999
```

Epic: CRDB-39093
Informs: #105340
Release note: None

cthumuluru-crdb committed:
fixup! all: generate tenant client certs using tenant names `cockroach cert create-tenant-client` command is used to generate client tenant certs. The generated cert file names are in `client-tenant.NN.crt` format where NN is the tenant ID.

cthumuluru-crdb commented on pkg/roachpb/test_test.go:
```diff
@@ -0,0 +1,30 @@
+package roachpb
```
Ignore this file. Committed by mistake and removed in fixup commit.
